plugins/factory-plugin/src/git.ts

```diff
@@ -0,0 +1,133 @@
+/*********************************************************************
+ * Copyright (c) 2019 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ **********************************************************************/
+const { spawn } = require('child_process');
+
+export interface GitBranch {
+
+    readonly isHEAD: boolean;
+
+    readonly localBranch: string;
+
+    readonly upstreamBranch?: {
+        remote: string,
+        branch: string,
+        remoteURL?: string
+    };
+
+    readonly aheadBehind?: {
+        ahead?: number,
+        behind?: number
+    };
+
+    /**
+     * The hash string of the current HEAD.
+     */
+    readonly ref: string;
+
+}
+
+export async function getCurrentBranch(projectPath: string): Promise<GitBranch | undefined> {
+
+    return new Promise<GitBranch | undefined>((resolve, reject) => {
+        const gitBranchCmd = spawn('git', ['branch', '-vv'], { cwd: projectPath });
+        const grep = spawn('grep', ['\*']);
+
+        gitBranchCmd.stdout.pipe(grep.stdin);
+
+        let output = '';
+        grep.stdout.on('data', function(data: string) {
+            output += data.toString();
+        });
+        grep.on('close', async function(code: any) {
+            output = output.trim();
+            const gitBranch: GitBranch | undefined = parseGitBranch(output);
+            if (gitBranch && gitBranch.upstreamBranch) {
+                gitBranch.upstreamBranch.remoteURL = await getRemoteURL(gitBranch.upstreamBranch.remote, projectPath);
+            }
+
+            resolve(gitBranch);
+        });
+        grep.on('error', function(err: any) { reject(err); });
+    });
+}
+
+export function getRemoteURL(remote: string, projectPath: string): Promise<string | undefined> {
```

Comment:
it looks like very fragile to parse all this stuff. Maybe I didn't understood what you're trying to get ?

Reply:
I am not interested in getting the local branch but the upstream branch as this would be what we would try to clone: upstream url + upstream branch.
Then, having covered all these parsing stuff with tests makes less fragile in my opinion.

Reply:
i will try `rev-parse`

Reply:
```bash
$ git rev-parse --abbrev-ref --symbolic-full-name @{u}
```

Reply:
(for getting upstream branch used locally)

Reply:
thanks :)